### DISCUSSION TITLE
Use of mutiple inputs for the neural network

### DISCUSSION THREAD

**Discussion by joaofernandez2011** [lululxvi/deepxde#1281] [Q&A]
Dear Lu, 
most probably my question is a rather easy question but I just recently started applying DeepXDE to my research. I am trying t develope a model for supercritical fluid extraction using physics informed neural networks.
It is clear to me that the dataset.train and dataset.test are lists of time/yield pairs of values. But for each pair a values there is a whole set of input information like (Pext, Text, Gext, time, Psize, q0, molecular mass of extracted compound, length of packed bed, diameter of extractor). My question is how to introduce this information in the model.

On a side question my model refuses to compile, below you will find the code:

alfa = 0.745
beta = 0.538
K = 50

def U(Gext, rho, diameter):
    return Gext / (rho * math.pi * (diameter /2) ** 2     # fluid phase velocity
                   
def Us(U, eps):
    return U / eps          # fluid phase superficial velocity

def h(Psize):
    return Psize/2      # characteristic dimension of the bed particle

def Bi(kf, h, De):
    return kf * h /De

def kp(kf, Bi):
    return kf / (1 + Bi/3)

def ap(h):
    return 1/h

def tau(width, lengt, Gext, rho):
    return math.pi * ((width/2)**2) * length / (Gext/rho)

def phi(kp, ap, tau):
    return kp * ap * tau

def theta(time):
    return time/tau

def Pe(length, Us, Dl):
    return length * Us / Dl

geom = dde.geometry.Rectangle(xmin=(0,0), xmax=(1,1))
timedomain = dde.geometry.TimeDomain(0, 1)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

def balances(theta, y):
    x, xs = y[:, 0:1], y[:, 1:2]
    dx_theta = dde.grad.jacobian(x, theta, i=0, j=1)
    dxs_theta = dde.grad.jacobian(xs, theta, i=1, j=1)
    Eq_a = dx_theta + x/alfa - (1-alfa)/alfa * (x - xs/K)
    Eq_b = dxs_theta - phi * (x -xs/K)/(beta/K + (1-beta))
    return [Eq_a, Eq_b]

# Initial conditions

def fun_init_x(x):
    return 0

def fun_init_xs(xs):
    return K / (beta + (1 - beta) * K)

ic_1 = dde.icbc.IC(geomtime, fun_init_x, lambda _, on_initial: on_initial, component=0)
ic_2 = dde.icbc.IC(geomtime, fun_init_xs, lambda _, on_initial: on_initial, component=1)

# Boundary conditions

def fun_bc(x):
    return 1 - x[:, 0:1]

bc_a = dde.icbc.DirichletBC(
    geomtime, fun_bc, lambda _, on_boundary: on_boundary, component=0
)
bc_b = dde.icbc.DirichletBC(
    geomtime, fun_bc, lambda _, on_boundary: on_boundary, component=1
)

data = dde.data.PDE(geomtime, balances, [ic_1, ic_2, bc_a, bc_b], num_domain=1200, num_boundary=120, num_test=1500)

layer_size = [1] + [50] * 3 + [1]
activation = "tanh"
initializer = "Glorot uniform"
net = dde.nn.FNN(layer_sizes=layer_size, activation=activation, kernel_initializer=initializer, regularization=None)

model = dde.Model(data, net)

model.compile("adam", lr=0.001, metrics=["l2 relative error"])
losshistory, train_state = model.train(iterations=2000)

**The error it gives is the following:**

Compiling model...

---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)

[<ipython-input-37-5d1ea8ec81ab>](https://localhost:8080/#) in <cell line: 3>()
      1 model = dde.Model(data, net)
      2 
----> 3 model.compile("adam", lr=0.001, metrics=["l2 relative error"])
      4 losshistory, train_state = model.train(iterations=2000)

9 frames

[/usr/local/lib/python3.10/dist-packages/deepxde/gradients.py](https://localhost:8080/#) in __call__(self, i, j)
     39             raise ValueError("i={} is not valid.".format(i))
     40         if j is not None and not 0 <= j < self.dim_x:
---> 41             raise ValueError("j={} is not valid.".format(j))
     42         # Compute J[i]
     43         if i not in self.J:

ValueError: j=1 is not valid.

I thank you any help or guidance that you may offer me.

**Comment by lululxvi** [comment]
https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant

**Comment by joaofernandez2011** [comment]
Dear Lu Lu, thank you very much for pointing the way. How can I introduce the exogeneous data in my model? I have an array with the form

    pext  text   gext  time   psize   q0      mm  length   width    rend
0    250    40  0.051     0  0.0026  2.6  222.37    0.16  0.0252  0.3766
1    250    40  0.051   600  0.0026  2.6  222.37    0.16  0.0252  0.8550
2    250    40  0.051  1200  0.0026  2.6  222.37    0.16  0.0252  1.2110
3    250    40  0.051  1800  0.0026  2.6  222.37    0.16  0.0252  1.5570
4    250    40  0.051  2400  0.0026  2.6  222.37    0.16  0.0252  1.8120
..   ...   ...    ...   ...     ...  ...     ...     ...     ...     ...
80   150    80  0.051  7200  0.0026  2.6  222.37    0.16  0.0252  0.7634
81   150    80  0.051  7800  0.0026  2.6  222.37    0.16  0.0252  0.8041
82   150    80  0.051  8400  0.0026  2.6  222.37    0.16  0.0252  0.8244
83   150    80  0.051  9000  0.0026  2.6  222.37    0.16  0.0252  0.8448
84   150    80  0.051  9600  0.0026  2.6  222.37    0.16  0.0252  0.8448

each line designates the combination of parameters for an experiment. And the data must be available to make interstep calculations.
Many thanks in advance,

Kind regards,
Joao Fernandes

**Reply by lululxvi** [reply]
https://deepxde.readthedocs.io/en/latest/demos/pinn_inverse/lorenz.inverse.forced.html

**Comment by joaofernandez2011** [comment]
Dear Lu Lu, thank you very much for answering. So if I understood correctly
I feed the points to deepXDE as an array. But my array has 11 dimensions:

theta, Yield, Pext, text, gext, time, psize, q0, mm, length, diameter

Is there a way to create continuous functions in so many dimensions (maybe
with a neural network)?

Lu Lu ***@***.***> escreveu no dia quinta, 18/05/2023 à(s)
23:24:

**Reply by lululxvi** [reply]
I cannot understand "Is there a way to create continuous functions in so many dimensions"
